Add utility to generate coverage map for Spark functions

zhejiangxiaomai opened the pull request:
- Add velox_sparksql_coverage CLI to generate coverage map for Spark functions.
- Refactor presto/coverage/Coverage.cpp to extract core logic into CoverageUtil
  class and re-use it to generate coverage maps for Spark functions.

mbasmanova commented:
> add handle_signature into spark.py. This function is necessary, the function link will fail without this function.

Can you move this fix into a separate PR?

mbasmanova commented on velox/functions/sparksql/coverage/Coverage.cpp:
```diff
@@ -0,0 +1,380 @@
+
+} // namespace
+
+int main(int argc, char** argv) {
```
Is this a copy-paste of the existing Coverage.cpp?  If so, let's refactor to avoid code duplication.

zhejiangxiaomai replied:
I will refactor to avoid code duplication.

mbasmanova commented:
@zhejiangxiaomai I'm also seeing nightly doc-gen job failing. Would you take a look?

https://app.circleci.com/pipelines/github/facebookincubator/velox/19730/workflows/7fffc2df-0393-4543-9788-e6abec45ff2b/jobs/124340

```
Exception occurred:
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/root/project/velox/docs/ext/spark.py", line 17
    from __future__ import annotations
                                     ^
SyntaxError: future feature annotations is not defined
```

zhejiangxiaomai commented:
> Exception occurred:
>   File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
>   File "/root/project/velox/docs/ext/spark.py", line 17
>     from __future__ import annotations
>                                      ^
> SyntaxError: future feature annotations is not defined

The solution of this issue to upgrade python version >=3.7.  `from __future__ import annotations` this code line is necessary for domain.

zhejiangxiaomai commented:
- Can you move this fix into a separate PR?
https://github.com/facebookincubator/velox/pull/3929

mbasmanova commented on velox/functions/CoverageUtil.h:
```diff
@@ -0,0 +1,316 @@
+#include <vector>
+#include "velox/functions/FunctionRegistry.h"
+
+using namespace facebook::velox;
```
Do not use "using" directive in header files. Move the implementation into .cpp file and leave only necessary declarations in the .h file.

zhejiangxiaomai replied:
Thanks for your comments. just fixed it.

mbasmanova commented on velox/functions/CoverageUtil.h:
```diff
@@ -0,0 +1,150 @@
+
+std::vector<std::string> readFunctionNamesFromFile(const std::string& filePath);
+
+std::string toFuncLink(const std::string& name, const std::string& domain = "");
```
Is this function used outside of the the CoverageUtil.cpp? If not, let's move it to the .cpp file. Please, go over all functions / classes in this .h file and move the ones that are not used outside of the .cpp file there.

zhejiangxiaomai replied:
Thanks, just fixed it. now the header file looks nicer.

mbasmanova commented on velox/functions/sparksql/coverage/Coverage.cpp:
```diff
@@ -0,0 +1,93 @@
+
+namespace {
+
+void printVeloxFunctions() {
```
What's the difference between this function and the one used in Presto's coverage?

zhejiangxiaomai replied:
Presto has a [kLinkBlockList](https://github.com/facebookincubator/velox/pull/3927/files#diff-6ab510aea06f78c8c4854b0d83bd7dafc70b20f651a41355a598456a2cbdaff8L291). And it is not useful for spark. May I can let kLinkBlockList as a arguments of printVeloxFunctions.

mbasmanova replied:
A block list can be an argument and Spark can pass an empty list.

zhejiangxiaomai replied:
Thanks, just fixed it.

mbasmanova commented on velox/functions/CoverageUtil.h:
```diff
@@ -0,0 +1,138 @@
+
+std::vector<std::string> readFunctionNamesFromFile(const std::string& filePath);
+
+class TablePrinter {
```
Can this class be moved into .cpp file?

zhejiangxiaomai replied:
Yes, you are right. The CoverageUtil.h only have three functions which are used outside.

mbasmanova commented on velox/functions/CoverageUtil.h:
```diff
@@ -0,0 +1,138 @@
+  std::vector<std::pair<int, int>> cells_;
+};
+
+void printCoverageMap(
```
Can this function go into .cpp file?

mbasmanova commented on velox/functions/CoverageUtil.h:
```diff
@@ -0,0 +1,138 @@
+void printCoverageMapForAll(const std::string& domain = "");
+
+void printVeloxFunctions(
+    const std::unordered_set<std::string> kLinkBlockList,
```
const std::unordered_set<...>& linkBlockList

zhejiangxiaomai replied:
Just fixed it.

mbasmanova commented on velox/functions/CoverageUtil.h:
```diff
@@ -0,0 +1,27 @@
+#include <string>
+#include <unordered_set>
+
+void printCoverageMapForAll(const std::string& domain = "");
```
Nice small header file. Thank you for cleaning it up. Let's document the remaining functions. In particular, let's clarify what "domain" is and explain that these functions expect to find certain files at certain paths.

zhejiangxiaomai replied:
Thanks, I added some documents for the remaining functions.

mbasmanova commented on velox/functions/CoverageUtil.h:
```diff
@@ -0,0 +1,27 @@
+void printCoverageMapForAll(const std::string& domain = "");
+
+void printVeloxFunctions(
+    const std::unordered_set<std::string>& kLinkBlockList,
```
kLinkBlockList -> linkBlockList

zhejiangxiaomai replied:
fixed.

mbasmanova commented on velox/functions/CoverageUtil.h:
```diff
@@ -0,0 +1,27 @@
+    const std::unordered_set<std::string>& kLinkBlockList,
+    const std::string& domain = "");
+
+void printCoverageMapForMostUsed();
```
Should this function take domain argument as well?

zhejiangxiaomai replied:
You are right.

mbasmanova commented on velox/functions/sparksql/coverage/Coverage.cpp:
```diff
@@ -0,0 +1,51 @@
+using namespace facebook::velox;
+
+namespace {
+// Do not add links for arithmetic operators.
```
"in" is not an arithmetic operator. Please, update. BTW, why not?

zhejiangxiaomai replied:
At first time, I just copied the logic of presto [here](https://github.com/facebookincubator/velox/blob/main/velox/functions/prestosql/coverage/Coverage.cpp#L291-L300).  In fact, I think this map can be empty for spark.

mbasmanova commented on velox/functions/sparksql/coverage/Coverage.cpp:
```diff
@@ -0,0 +1,51 @@
+
+namespace {
+// Do not add links for arithmetic operators.
+static const std::unordered_set<std::string> kLinkBlockList = {"in"};
```
consider moving this constant into main, right before it is used

zhejiangxiaomai replied:
Just fixed it.

mbasmanova commented on velox/functions/sparksql/coverage/README.md:
```diff
@@ -0,0 +1,23 @@
+
+> velox_sparksql_coverage --all
+
+Generates coverage map using all Spark functions. The output to be copy-pasted
```
It would be nice to automate this process and add it to the nightly doc-gen job. CC: @kgpai

kgpai replied:
Hi Masha, Sure I will do that; Also As I pointed out in a comment above, it should be possible to have this generated via sphinx code automatically without the need for a separate process via pyVelox. It doesnt have to be done in this PR, but we should aim to move such utility code to PyVelox.

mbasmanova reviewed (commented):
Thank you for the updates. Some further comments. Would you also attach screenshots of the generated docs?

mbasmanova commented on velox/docs/spark_functions.rst:
```diff
@@ -17,4 +17,67 @@ Spark Functions
     functions/spark/aggregate
 
+Here is a list of all scalar and aggregate Spark functions available in Velox.
+Function names link to function descriptions. Check out coverage maps
```
typos: "Check out coverage maps for all ." -> "... map for all functions."

mbasmanova commented on velox/functions/sparksql/coverage/Coverage.cpp:
```diff
@@ -0,0 +1,46 @@
+  if (FLAGS_all) {
+    printCoverageMapForAll(":spark");
+  } else {
+    static const std::unordered_set<std::string> linkBlockList;
```
This variable is not needed. Just pass "".

mbasmanova replied:
For some reason, I see "in" without a link in the docs. Would you re-generate?

mbasmanova replied:
Also, links do not work for me for some reason. Do these work for you?

zhejiangxiaomai replied:
> This variable is not needed. Just pass "".
It seems that "" can not pass compile.
../../velox/functions/sparksql/coverage/Coverage.cpp:41:25: error: invalid initialization of reference of type ‘const std::unordered_set<std::__cxx11::basic_string<char> >&’ from expression of type ‘const char [1]’
   41 |     printVeloxFunctions("", ":spark");

zhejiangxiaomai replied:
> Also, links do not work for me for some reason. Do these work for you?

I Also feel curious. I think some change in this pr cause link fail. because I found  link can work with velox latest main branch.

mbasmanova replied:
> It seems that "" can not pass compile.

Oh. Sure, pass {}.

```
printVeloxFunctions({}, ":spark");
```

> I Also feel curious. I think some changes in this pr cause link fail. because I found link can work with velox latest main branch.

Would you investigate?

mbasmanova commented:
Seeing some errors:

/Users/mbasmanova/cpp/velox-1/velox/docs/functions/spark/math.rst:23: ERROR: Unexpected indentation.
/Users/mbasmanova/cpp/velox-1/velox/docs/functions/spark/string.rst:77: ERROR: Unexpected indentation.

zhejiangxiaomai commented:
1. Link can work now.
2. fix "ERROR: Unexpected indentation.".

mbasmanova commented:
@zhejiangxiaomai 

I'm seeing some warnings still:

```
/Users/mbasmanova/cpp/velox-1/velox/docs/functions/spark/string.rst:49: WARNING: duplicate object description of split, other instance in functions/spark/string, use :noindex: for one of them
/Users/mbasmanova/cpp/velox-1/velox/docs/functions/spark/string.rst:71: WARNING: duplicate object description of substring, other instance in functions/spark/string, use :noindex: for one of them
```

I also see some links in the coverage map working (instr, pmod), but some don't (aggregate, array, array_contains, least, rand). Would you double check?

zhejiangxiaomai commented:
-- I also see some links in the coverage map working (instr, pmod), but some don't (aggregate, array, array_contains, least, rand). Would you double check?

I really appreciate your opinion. All functions were linked now and fixed all warnings about spark.  And I removed linkBlockList.

mbasmanova commented:
@zhejiangxiaomai I still see some links not working, e.g. least, greatest. Do all links work for you?

I see you added isnotnull and isnull functions to Array Functions doc. Do they belong there? Shouldn't they go into comparison.rst instead?

I also see documentation for the aggregate function, but do we have it implemented? Would you double check?

zhejiangxiaomai commented:
Hi,  @mbasmanova. 
1. All links in spark_functions.html can work.  I removed some outdate functions(like getarrayitem,getmapvalue,murmur3hash) which are not presented in https://spark.apache.org/docs/latest/api/sql/.  And add documents for some functions(like unaryminus).
2. move isnotnull and isnull to comparison.html.
3.  aggragte function can work. here is a screen shot from spark+velox.
![image](https://user-images.githubusercontent.com/14065538/218358431-550eee37-13e5-4634-8686-805df5f07169.png)

mbasmanova commented on velox/functions/sparksql/Register.cpp:
```diff
@@ -97,18 +97,12 @@ void registerFunctions(const std::string& prefix) {
       prefix + "rlike", re2SearchSignatures(), makeRLike);
   VELOX_REGISTER_VECTOR_FUNCTION(udf_regexp_split, prefix + "split");
 
-  // Subscript operators. See ExtractValue in complexTypeExtractors.scala.
```
Let's mention these changes in the PR description. Or, better yet, extract these into a separate PR.

zhejiangxiaomai replied:
I extracted these into https://github.com/facebookincubator/velox/pull/4095.

mbasmanova commented on velox/docs/functions/spark/array.rst:
```diff
@@ -2,6 +2,31 @@
 Array Functions
 =============================
 
+.. spark:function:: aggregate(array(E), start, merge, finish) -> array(E)
```
Do we have an implementation for this function? Is it even scalar function? Would you please double check?

zhejiangxiaomai replied:
https://github.com/facebookincubator/velox/blob/main/velox/functions/sparksql/Register.cpp#L44
Yes, function aggregate was registered here.

mbasmanova commented on velox/docs/functions/spark/string.rst:
```diff
@@ -69,12 +76,28 @@ Unless specified otherwise, all functions return NULL if at least one of the arg
\ No newline at end of file
+        SELECT substring('Spark SQL', 5, 10000); -- "k SQL"
+
+.. spark:function:: replace(string, search, replace) -> string
```
Functions should appear in alphabetical order. Please, move.

zhejiangxiaomai replied:
Just fixed it.

mbasmanova commented on velox/functions/sparksql/coverage/Coverage.cpp:
```diff
@@ -0,0 +1,45 @@
+ */
+
+#include <gflags/gflags.h>
+#include <fstream>
```
Do we need all these includes?

zhejiangxiaomai replied:
Just removed it.

mbasmanova commented on velox/functions/sparksql/tests/SubscriptTest.cpp:
```diff
@@ -41,42 +41,4 @@ class SubscriptTest : public SparkFunctionBaseTest {
 
 } // namespace
-
-// Spark's subscript ("a[1]") behavior:
```
Why are you deleting this test?

zhejiangxiaomai replied:
Because this UT contains outdate function "getarrayitem" and "getmapvalue". And I just removed those two functions in https://github.com/facebookincubator/velox/pull/4095

zhejiangxiaomai replied:
Hi, @mbasmanova would you take a look at https://github.com/facebookincubator/velox/pull/4095.

mbasmanova replied:
@zhejiangxiaomai I just looked at #4095 and left some comments. Thank you for extracting these changes into a separate PR.

mbasmanova reviewed (commented):
@zhejiangxiaomai Thank you for iterating on this PR. It is shaping up nicely. I still see some missing links in html/spark_functions.html. For example, bitwise_and, bitwise_or, get_json_object, hash, md5, shiftleft, shiftright, xxhash64, year. Also, I don't see a link to coverage map.

zhejiangxiaomai commented:
-- I still see some missing links in html/spark_functions.html. 
I guess you need pull my latest branch. or run "make clean". Because I double checked it that all function link can work. here is a screen shot. 
![image](https://user-images.githubusercontent.com/14065538/220041829-d425088b-dc4a-4acc-9d2c-840e2b0f441e.png)

--Also, I don't see a link to coverage map.
I found that coverage.html at presto part also do not have link. So I just follow this way.  Do you mean we need add spark/coverage.html into index.html?

kgpai commented on velox/docs/functions/spark/aggregate.rst:
```diff
@@ -12,6 +12,7 @@ General Aggregate Functions
 
 .. spark:function:: last(x, ignoreNull) -> x
-
+    :noindex:
```
why is there noindex on certian functions ?

zhejiangxiaomai replied:
Thanks, just remove it.

kgpai commented on velox/docs/functions/spark/map.rst:
```diff
@@ -13,12 +13,26 @@ Map Functions
+
+.. spark:function:: map_from_arrays(array(K), array(V)) -> map(K,V)
+
+    Creates a map with a pair of the given key/value arrays. All elements in keys should not be null. ::
```
nit: Can you add documentation if key size != value size , what are possible outcomes ?

zhejiangxiaomai replied:
just added it. If key size != value size will throw exception that key and value must have the same length.

kgpai commented on velox/functions/CoverageUtil.cpp:
```diff
@@ -0,0 +1,387 @@
+ * limitations under the License.
+ */
+
+#include "velox/functions/CoverageUtil.h"
```
nit: I think it should be possible to move all of these to pyVelox ; This way it can be imported into Sphinx and generated from inside sphinx without the need to run as a separate function. 
FWIW we can do this in a separate PR though in the future.

kgpai commented on velox/functions/sparksql/coverage/Coverage.cpp:
```diff
@@ -0,0 +1,44 @@
+ */
+
+#include <gflags/gflags.h>
+#include <iostream>
```
nit: is iostream required here?

zhejiangxiaomai replied:
just remove it.

kgpai reviewed (commented):
Looks very good @zhejiangxiaomai ! Thanks for your work on this.

mbasmanova commented:
@zhejiangxiaomai 

> -- I still see some missing links in html/spark_functions.html.
I guess you need pull my latest branch. or run "make clean". Because I double checked it that all function link can work. here is a screen shot.

I tried again and still do not see all the links. I noticed that some files render incorrectly, see example attached. Notice how "year(x) -> integer" is all bold. Only "year" should be bold. I edited the file to add a space, ran make html and the issue got fixed. I removed the space, re-ran make html and the issue stayed fixed. I wonder if some files have some weird characters. We had a similar issue some time back. At that time it was because the files were created on Windows with some incorrect settings. Would you please double check?

<img width="482" alt="Screenshot 2023-02-27 at 11 02 51 AM" src="https://user-images.githubusercontent.com/27965151/221615401-99328d77-8861-4752-9d6b-1f843796aa9c.png">

mbasmanova commented on velox/functions/CMakeLists.txt:
```diff
@@ -11,7 +11,7 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
-add_library(velox_function_registry FunctionRegistry.cpp)
+add_library(velox_function_registry FunctionRegistry.cpp CoverageUtil.cpp)
```
That's a good point. Let's move CoverageUtil into a separate library.

zhejiangxiaomai replied:
Ok, I will fix comments tomorrow.

zhejiangxiaomai replied:
I move CoverageUtil into a separate library.

zhejiangxiaomai commented:
@kgpai Just fixed all comments.  would you review this PR again when you are free?

mbasmanova commented on velox/docs/functions/spark/aggregate.rst:
```diff
@@ -12,6 +12,6 @@ General Aggregate Functions
 
 .. spark:function:: last(x, ignoreNull) -> x
-
+    
```
Unintended change?

zhejiangxiaomai replied:
fixed it.

mbasmanova commented on velox/docs/functions/spark/array.rst:
```diff
@@ -33,7 +72,7 @@ Array Functions
 
 .. spark:function:: sort_array(array(E), ascendingOrder) -> array(E)
-
+    
```
same

zhejiangxiaomai replied:
fixed it.

mbasmanova commented on velox/docs/functions/spark/struct.rst:
```diff
@@ -0,0 +1,9 @@
+Struct Functions
+====================================
+
+.. spark:function:: named_struct(name1, val1, name2, val2, ..., namen, valn) -> struct
```
I'm not sure this "function" takes arguments like this. I also don't even think this works at all as I see logic in velox/expression/ExprCompiler.cpp that works only for "row_constructor".

```
const char* const kRowConstructor = "row_constructor";
...
ExprPtr getSpecialForm(
    const std::string& name,
    const TypePtr& type,
    std::vector<ExprPtr>&& compiledChildren,
    bool trackCpuUsage) {
  if (name == kRowConstructor) {
    return getRowConstructorExpr(
        type, std::move(compiledChildren), trackCpuUsage);
  }
```

It would be nice to add a unit test to confirm that "named_struct" works.

zhejiangxiaomai replied:
Ok. I will double check it.

zhejiangxiaomai replied:
Yes, you are right. velox just register [here](https://github.com/facebookincubator/velox/blob/main/velox/functions/sparksql/Register.cpp#L56). But UT can not pass. I removed this named_struct document.

mbasmanova replied:
@zhejiangxiaomai Thank you for checking. Let's mention that named_struct function was removed in the PR description.

mbasmanova reviewed (commented):
@zhejiangxiaomai Thank you for iterating on this PR. Looks good overall % a couple comments. Let's rebase to pick up recent changes to coverage tool that added support for window functions.

mbasmanova commented on velox/functions/sparksql/Register.cpp:
```diff
@@ -53,7 +53,7 @@ static void workAroundRegistrationMacro(const std::string& prefix) {
   // This is the semantics of spark.sql.ansi.enabled = false.
   VELOX_REGISTER_VECTOR_FUNCTION(udf_element_at, prefix + "element_at");
-  VELOX_REGISTER_VECTOR_FUNCTION(udf_concat_row, prefix + "named_struct");
+  // VELOX_REGISTER_VECTOR_FUNCTION(udf_concat_row, prefix + "named_struct");
```
Let's remove this code, not comment out.

mbasmanova reviewed (approved):
@zhejiangxiaomai Looks great % a couple of nits.
